lib/Dialect/FIRRTL/Transforms/InjectDUTHierarchy.cpp

```diff
@@ -0,0 +1,230 @@
+//===- InjectDUTHierarchy.cpp - Add hierarchy above the DUT ---------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of the SiFive transform InjectDUTHierarchy.  This moves all
+// the logic inside the DUT into a new module named using an annotation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/NLATable.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct InjectDUTHierarchy : public InjectDUTHierarchyBase<InjectDUTHierarchy> {
+
+  /// The design-under-test (DUT).  This is kept up-to-date by the pass as the
+  /// DUT changes due to internal logic.
+  FModuleOp dut;
+
+  /// The wrapper module that is created inside the DUT to house all its logic.
+  FModuleOp wrapper;
+
+  /// The name of the new module to create under the DUT.
+  StringAttr wrapperName;
+
+  /// Mutable indicator that an error occurred for some reason.  If this is ever
+  /// true, then the pass can just signalPassFailure.
+  bool error = false;
+
+  void runOnOperation() override;
+};
+} // namespace
+
+void InjectDUTHierarchy::runOnOperation() {
+
+  CircuitOp circuit = getOperation();
+
+  AnnotationSet::removeAnnotations(circuit, [&](Annotation anno) {
+    if (!anno.isClass(injectDUTHierarchyAnnoClass))
```

Comment:
you might want to check if we have already errored and stop looking for more errors
```c++
if (error)
  return false;
```